feat(node): v11.6

SimonSchick opened the pull request:
Please fill in this template.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V11.md
- [x] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

This adds misc things from node v11.6, some of the changes are being worked on by others so I didn't add those, see:

In addition I've started moving module tests into the /tests folder to avoid constantly having to touch this overly large file.

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/32787
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/32835
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/32597

@rbuckton as a note, would you be fine with removing typings for all node version that are EOL? (v0, v4, v5, v7, v9) it should prevent people from editing old/obsolete typings.

Also:
fixes #32744
fixes #32926
Partially addresses #31505

SimonSchick commented:
@Flarna knowing the @types/node CI this will take a while to pass, feel free to leave comments :)

SimonSchick commented:
Ok, having some problems with `TextEncoder`, mostly packages loading dom and node typings.

Flarna commented:
I think it's a good idea to remove obsolete type definitions but I would do this in a separate PR just in case someone complains and a revert is needed. CI for node was green yesterday so maybe a good time window today.

SimonSchick commented:
@Flarna problem is that `TextEncoder` conflicts with dom typings now, I don't see how I can fix this :/

Flarna commented:
I think #25342 was about a similar problem with `URL` which was added to `global`. Solution that time was to simply skip this and it seems no one complained till now :o).

In general I recommend to keep using module `dom` in tests here as any incompatibilities with `dom` result in a lot broken builds in the wild.

SimonSchick commented:
@Flarna yea that's what I thought too, problem is that this really isn't a sustainable solution.

SimonSchick commented:
Ok, reverted `TextEncoder` changed to make sure this can be merged.

In addition I went over the deprecation log, added deprecation tags and removed dead APIs as appropriate.

SimonSchick commented:
@armanio123 unrelated build failure due to:
```
Error in s3-download-stream
Error: /home/travis/build/DefinitelyTyped/DefinitelyTyped/types/s3-download-stream: Contains the word 'download', which is banned by npm.
```
I don't think this is my fault, should I remove the package or something?

armanio123 commented:
@SimonSchick it seems there's a new rule added by npm that's breaking the types-publisher. We are disallowing it for the moment while we have a better solution. @sandersn have a more context on this.

SimonSchick commented:
@sandersn can we skip/warn on those packages instead of failing the build if it's not the primary package being tested? 😞

SimonSchick commented:
Also can you provide the source on this rule change, can't find anything on this.

sandersn commented:
Here's what I got from NPM support this morning:

> The word `download` is not allowed in package names so you will have to find an alternative for your package. Unfortunately, we are unable to override this.

Then, after asking for the banned word list so I could avoid trying to publish packages with those words:
> Unfortunately, we don't currently have a public list of banned words. You will see "That word is not allowed" when a word doesn't pass the spam algorithm.

This is the only npm-official reference I found to spam filtering: https://blog.npmjs.org/post/169582189317/incident-report-npm-inc-operations-incident-of. I assume they don't talk much about it because they're going for security by obscurity, or are not completely in control of their filtering software and don't know ahead of time what will be rejected.


There are only 3 packages that currently use the word 'download': 's3-download-stream', 'download' and 'downloadjs'. I assume that those packages are allowed just because they were on NPM before the rule change. I'll manually exclude them from the rule and update dtslint, because I *really* don't like random packages blocking `@types` publishing.

SimonSchick commented:
That's very un-transparent of them, are there any plans to create a secondary package registry (run by DT/MS) and map an org (for examples @types2) to that registry?

Anyways, is there a tracking issue for this, I'd like to voice my opinion there!

sandersn commented:
I pushed a fix to dtslint, so I'll restart the Travis build.

SimonSchick commented:
👍

SimonSchick commented:
Also went ahead and:
fixed #32926 in this major version as it could be considered a breaking change.
Added `SharedArrayBuffer` as class.
Enabled strict null checks.

Flarna commented:
#32661 and #32661 have been merge recently and therefore these changes are missing in our v10 variant.

SimonSchick commented:
@Flarna you posted the same PR twice.
Also I think I can just rebase?

Flarna commented:
Ups.. corrected comment above. I don't think rebase is enough as this would not update the files in v10 folder because they don't exist on master. Git is usually not able to track something like this.

SimonSchick commented:
done

Flarna commented on types/node/child_process.d.ts:
```diff
@@ -77,32 +77,34 @@ declare module "child_process" {
-    interface ExecOptions {
-        cwd?: string;
+    interface ExecOptions extends CommonOptions {
         env?: NodeJS.ProcessEnv;
```
`env` is already included via `CommonOptions`

Flarna commented on types/node/child_process.d.ts:
```diff
@@ -333,13 +320,11 @@ declare module "child_process" {
         input?: string | Buffer | NodeJS.TypedArray | DataView;
         stdio?: StdioOptions;
         env?: NodeJS.ProcessEnv;
-        uid?: number;
```
Is `extends CommonOptions` missing here? if yes, `env`, `cwd` are already included there.

SimonSchick replied:
No because of `windowsHide`, I tried to dedupe a little bit here and wanted to avoid creating too many `common` interfaces.

Flarna replied:
Sorry, don't get this. [docs](https://nodejs.org/dist/latest-v11.x/docs/api/child_process.html#child_process_child_process_execfilesync_file_args_options) tell that `uid`, `env`, and `windowsHide` are valid options for `ExecFileSync`

Flarna commented on types/node/crypto.d.ts:
```diff
@@ -164,7 +154,13 @@ declare module "crypto" {
         getPublicKey(encoding: HexBase64Latin1Encoding): string;
         getPrivateKey(): Buffer;
         getPrivateKey(encoding: HexBase64Latin1Encoding): string;
+        /**
```
according to docs `DiffieHellman.setPrivateKey` and `DiffieHellman.setPublicKey` are not deprecated but `ecdh.setPublicKey` is.

SimonSchick replied:
Oh, we don't have that, so I'm just going to ignore it.

Flarna commented on types/node/fs.d.ts:
```diff
@@ -275,6 +275,7 @@ declare module "fs" {
     /**
      * Asynchronous lchown(2) - Change ownership of a file. Does not dereference symbolic links.
      * @param path A path to a file. If a URL is provided, it must use the `file:` protocol.
+     * @deprecated
```
according to [docs](https://nodejs.org/dist/latest-v11.x/docs/api/fs.html#fs_fs_lchown_path_uid_gid_callback) `lchown` and `lchownSync` are no longer deprecated since 10.6.0

Flarna commented on types/node/tty.d.ts:
```diff
@@ -7,7 +7,36 @@ declare module "tty" {
+        clearLine(dir: Direction): void;
+        clearScreenDown(): void;
+        cursorTo(x: number, y: number): void;
+        getColorDepth(env: {}): number;
```
```suggestion
        getColorDepth(env?: {}): number;
```

Flarna commented on types/node/index.d.ts:
```diff
@@ -1,4 +1,4 @@
-// Type definitions for Node.js 10.12
+// Type definitions for Node.js 11.6
```
```suggestion
// Type definitions for Node.js 11.8
```
`timeout` option in `tls.connect()` was added in 11.8

Flarna commented on types/node/ts3.1/tsconfig.json:
```diff
@@ -7,12 +7,11 @@
         "lib": [
-            "es6",
-            "dom"
+            "es6"
```
shouldn't we keep "dom" here to ensure that node and dom are compatible?

SimonSchick replied:
Yes and no.
I don't think they are going to be compatible in the long run and it's already causing issues (see TextEncoder/console etc.), eventually these will have to be broken up or `console` etc. need to be moved to a shared lib.

I will restore it for now but I still think it's a nasty anti pattern.

Flarna replied:
I' assume that a lot people don't do this on purpose; it just happens because of dependencies.

SimonSchick commented:
@rbuckton if you decide to merge this, please squash :)

SimonSchick commented:
@Flarna you happy? :P

SimonSchick commented:
@btoueg @Flarna merged changes from https://github.com/DefinitelyTyped/DefinitelyTyped/pull/32597 into this PR.
Also:
Cleaned up crypto file by introducing `Binary` and `BinaryLike` and `KeyLike` types to reduce type union bloat.

Flarna commented:
Merge seems to be ok. I have no special opinion regarding the new crypto types.
But it seems the removal of `crypto.Credentials` results in an unhappy irc module.

Any idea why other node maintainer have not been invited for a review by @typescript-bot ?

SimonSchick commented:
@Flarna according to the docs crypto.Credentials was wrong, unfortunately I can't run the node depedency tests locally, will fix soon. :)

qm3ster commented:
Thank you for the `toJSON(): { type: 'Buffer', data: number[] };`
Shouldn't the `encoding` argument in `Buffer.write` be the `BufferEncoding` union declared above instead of `string`? (I mentioned this as an also in #32926, not sure if anyone saw)

SimonSchick commented:
@qm3ster I guess so, would also make sense to create separate overloads for all other functions that take `encoding` as it's only required for strings.
Please create a new issue for this, not sure when I will get around to do it.
